Add monitoring of the servers time drifting

vapopov opened the pull request:
In this PR added snapshotting the local time for the services like: DB, Kube, Node, App by heartbeat functionality in case if these services are running in separate nodes with time difference bigger than 1 minute. We have monitor which periodically track the time difference and publish this information to global notification.

Related: https://github.com/gravitational/teleport/issues/22166
changelog: Teleport monitors time drift of remote servers to warn about unsynchronized time

```bash
root@3e255c82fc04:/go/teleport# ./build/tctl notifications ls --all
ID                                   Created             Expires             Title                                          Labels
------------------------------------ ------------------- ------------------- ---------------------------------------------- ------
cluster-monitor-system-clock-warning 28 Nov 24 01:28 UTC 28 Dec 24 01:28 UTC Incorrect system clock detected in the cluster
```

<img width="488" alt="image" src="https://github.com/user-attachments/assets/2d67d08a-caf4-4bc7-96ac-0e48fb0f6968">
<img width="1512" alt="image" src="https://github.com/user-attachments/assets/4628d485-e4ea-49bb-a56a-b94969ee6975">

vapopov committed:
Add monitoring of the servers time drifting

tigrato commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -233,6 +233,12 @@ message DatabaseServerSpecV3 {
   // ProxyIDs is a list of proxy IDs this server is expected to be connected to.
   repeated string ProxyIDs = 13 [(gogoproto.jsontag) = "proxy_ids,omitempty"];
+  // LocalTime stores information about the server local time.
+  google.protobuf.Timestamp LocalTime = 14 [
```
Should we refer to it as local time?

Local time represents the time in the machine's timezone, whereas `google.protobuf.Timestamp` does not include timezone information and is always in UTC.

It's important to note that the `*ServerSpec*` is not always updated or sent by the server. Our heartbeat mechanism prioritizes the KeepAlive (TTL extension at the database level) instead of sending the entire `*ServerSpec*` along with the upsert operation. As a result, the local time will reflect the time of the last change in the `*ServerSpec*` message.

However, if you use the watcher mentioned above, you will receive updates to `*ServerSpec*` each time the TTL key is extended, and this must be taken into account.

My suggestion is to use a separate process for time reconciliation, either through the `InventoryControlStream` or a different API, rather than relying on heartbeats. Using heartbeats could potentially cause issues for teleport.

vapopov replied:
Thanks @tigrato, will check this one

vapopov replied:
Was prototyping with inventory stream, tried to add local time as part of the metadata then after testing realized that is also cached data and refreshed only after Teleport restart, and time check should be similar to ping request, so replacing this one

tigrato replied:
This might be better clarified by @rosstimothy, but we can probably use the `Ping/Pong` or similar messages.

We only need the time difference if there is a drift that satisfies the notification criteria, we do not need to store it anywhere. 
In this case, I think we can use [any clock synchronization algorithm](https://en.wikipedia.org/wiki/Network_Time_Protocol#Clock_synchronization_algorithm) will satisfy our needs. We can "measure" the time drift between two clocks which is what we need.

Since we want to measure the time difference between the auth and the agent, the auth should periodically (eg every 24 hours) issue the time sync requests and measure the difference.

vapopov replied:
I've added additional command handling for time reconciliation, was not sure about ping/pong extension of new logic, just didn't want to mix responsibilities. We discussed about possible solution with making request to some predefined list of NTP servers, but this doesn't cover the case of blocked outbound traffic on the auth node.

tigrato replied:
I think using the `UpstreamInventoryPong`  and  `DownstreamInventoryPing` is a better fit given that all the logic is already present and we just need to add a new field to it and change the comment. This allow us to skip cases where auth supports this feature but client's don't - `XXX_unrecognized` is set

I don't think we need an ntp server for this. The issue happens when auth clock drifts from the server(s). If they are both wrong but with the same value, there won't be any issue accessing the servers. 

Sure it would be great to compare them against a real cock, but it's not a hard requirement to make teleport work correctly.

vapopov replied:
> This allow us to skip cases where  auth supports this feature but client's don't

we anyway need to add `google.protobuf.Timestamp LocalTime` to pong response in such case, so both auth and node should have these updates

tigrato replied:
Here is a rephrased version:

There is a difference between not recognizing a oneof field and not knowing that a particular message field exists. 

For example, if auth sends a clock request, it expects a response from the server. If the server does not recognize the message type, it will ignore it, as seen here: https://github.com/gravitational/teleport/pull/44489/files#diff-13c41552d21a59b707b926d991bc325421f323363b3c6e802085f4068ed8197bR495-R497

Because it ignores the message, it will not respond. However, if you use Ping/Pong, and the response timestamp is empty, it indicates that the server does not support the clock check and auth can safely continue

vapopov committed:
Merge remote-tracking branch 'origin/master' into vapopov/heartbeat-local-time-snap

vapopov commented on lib/auth/server_monitor.go:
```diff
@@ -0,0 +1,169 @@
+				duration := time.Duration(localTime - now)
+
+				if (duration > 0 && duration > timeShiftThreshold) || (duration < 0 && -duration > timeShiftThreshold) {
+					resources[res.GetName()] = resourceMonitor{resource: res, diff: duration}
```
yes, we discussed about this with Roman, initially since original issue related to certificates I've set auth service as point of truth. I will change this one to sorting times and informing the difference in general between services

r0mant replied:
Yeah I think we need to be smarter about how we create notifications - maybe we could just create a single notification saying something like "auth server X has time drift of Y compared to nodes node-1, node-2, node-3 and N others", something like that?

We definitely don't want to bombard the user with notifications for each server.

vapopov committed:
Refactor monitor to use inventory stream

vapopov committed:
Add command for inventory control stream for time reconciliation

vapopov committed:
Reuse ping/pong logic for time reconciliation

vapopov committed:
Add integration test to check if global notification is created

vapopov committed:
Merge remote-tracking branch 'origin/master' into vapopov/heartbeat-local-time-snap

vapopov committed:
Limits notification length, renaming

tigrato commented on lib/inventory/controller.go:
```diff
@@ -478,17 +493,21 @@ func (c *Controller) handlePong(handle *upstreamHandle, msg proto.UpstreamInvent
+		reqDuration: c.clock.Since(pending.start),
+	}
+	if !msg.SystemClock.IsZero() {
+		rsp.clockDiff = c.clock.Since(msg.SystemClock)
```
Instead of passing the since here, can we pass the real clock diff (this value less half of the req duration) because that's the real clock diff. 
This diff includes half of the latency and you don't need to fix it later if we fix it here.

If you want to have the logic in another place, pass the system time instead of clock diff to make it clear

tigrato commented on lib/inventory/inventory.go:
```diff
@@ -614,7 +636,35 @@ func (h *upstreamHandle) Ping(ctx context.Context, id uint64) (d time.Duration,
+		case rsp.clockDiff == 0:
+			return 0, nil
+		case rsp.clockDiff > 0:
+			return rsp.clockDiff - rsp.reqDuration/2, nil
```
Can we have the system times instead of duration req start time, req end time and downstreamSystemTime, we apply the formula and compute the absolute value after? 

That approach seems clear to me if we aren't interested in which service is behind in time

tigrato commented on lib/inventory/inventory.go:
```diff
@@ -614,7 +636,35 @@ func (h *upstreamHandle) Ping(ctx context.Context, id uint64) (d time.Duration,
+	case rsp := <-rspC:
+		switch {
+		case rsp.clockDiff == 0:
+			return 0, nil
```
Instead of returning zero here, let's return an error that indicates it's not supported as clock diffs can also be zero. 

The caller can then errors.Is to ignore the scenario

tigrato commented on lib/auth/server_monitor.go:
```diff
@@ -0,0 +1,148 @@
+			var inventories []string
+			a.inventory.Iter(func(handle inventory.UpstreamHandle) {
+				info := handle.Hello()
+				diff, err := handle.TimeReconciliation(ctx, insecurerand.Uint64())
```
Instead of running these in sequencial order, can we have some kind of parallelism?

espadolini replied:
Use `math/rand/v2` instead of `math/rand`.

tigrato commented on lib/inventory/inventory.go:
```diff
@@ -614,7 +636,35 @@ func (h *upstreamHandle) Ping(ctx context.Context, id uint64) (d time.Duration,
+		case rsp.clockDiff > 0:
+			return rsp.clockDiff - rsp.reqDuration/2, nil
+		default:
+			return rsp.clockDiff + rsp.reqDuration/2, nil
```
This formula seems incorrect. If the time difference is negative, we shouldn't add half of the req duration otherwise we are reducing the time difference.

espadolini commented on lib/auth/server_monitor.go:
```diff
@@ -0,0 +1,148 @@
+			return nil
+		case <-checkInterval.Next():
+			var inventories []string
+			a.inventory.Iter(func(handle inventory.UpstreamHandle) {
```
I don't think you're ever intended to use `(*inventory.Controller).Iter` with any visitor function that can block - it locks each shard of the inventory in order, and any new connection to the cluster will have to wait until the whole shard has been iterated through.

vapopov commented on lib/auth/server_monitor.go:
```diff
@@ -0,0 +1,148 @@
+// generateNotificationMessage formats the notification message with the inventory list.
+func generateNotificationMessage(messages []string) string {
+	var more string
+	if len(messages) > 10 {
+		more = fmt.Sprintf("Only 10 servers are shown out of the list of %d.", len(messages))
+		messages = messages[:10]
+	}
```
inventory iterator doesn't support limit, so we have to truncate or add the logic to inventory store

espadolini replied:
Does it really make sense to monitor servers periodically? Are we actually concerned about _drift_ rather than just misconfigurations?

And if we are, why an "external" periodic process rather than something that runs as part of the handling of the inventory control stream?

vapopov replied:
Generally periodically it runs to gather info about all nodes in cluster to generate one report with the list of nodes. I believe in most cases original issue related to misconfiguration, but there are other cases as well like this one with VM https://github.com/gravitational/teleport/issues/22166#issuecomment-2042862626 , another might be with the node in private network without local ntp server (not sure how much time needs for RTC to reach this 1min threshold inaccuracy)

@espadolini do you mean it should be a part of the inventory control stream (like periodic check and generating the list of nodes with time inconsistency and api to fetch this list)?

espadolini replied:
If this was part of the control stream handling you could just have the measured drift (if available) as part of the `upstreamHandle` and potentially added to the instance heartbeats, and any periodic report operation would just iterate the store and copy off the data, and then decide what to do about the report.

vapopov committed:
Return system clock instead of time difference
Add parallel requests to ping the nodes
Fix issue with locking inventory store while making ping request

vapopov committed:
Refactor to make time reconciliation part of the inventory control stream

vapopov committed:
Merge remote-tracking branch 'origin/master' into vapopov/heartbeat-local-time-snap

vapopov committed:
Fix tests

espadolini commented on lib/inventory/controller.go:
```diff
@@ -418,6 +457,31 @@ func (c *Controller) handleControlStream(handle *upstreamHandle) {
+			return
+		case pong := <-rspC:
+			tracker.mu.Lock()
+			tracker.systemClockDifference = c.clock.Since(pong.systemClock) - pong.reqDuration/2
```
What's the point of checking the time difference again here? Don't we have the ping start time and the "duration" already?

vapopov replied:
not sure if I understand correctly, do we need to store in state tracker time when request is send, response clock and request duration?

because when we request `timeReconciliation` from the `handleControlStream` loop, we make ping request to the handler receive system clock with some response delay, then store the time difference in state tracker and then in next heartbeat send this duration as part of `types.Instance`

espadolini replied:
We should measure our local time as close to the receive as possible, so I think it would be better to save local time, remote time (if provided) and duration in the `pingResponse` object.

vapopov replied:
@espadolini could you please take another look, if I do right inventory collection iteration. I've also just replaced in proto to expose clock data instead time difference duration

r0mant reviewed (commented):
@vapopov Are you still working on this? Can we address the latest round of feedback and ship this?

vapopov commented:
@r0mant yes, was planning to finish this week

vapopov committed:
Rewrite tests
Return inventory store iterator

vapopov committed:
Merge remote-tracking branch 'origin/master' into vapopov/heartbeat-local-time-snap

vapopov committed:
Naming adjustments

vapopov committed:
Merge remote-tracking branch 'origin/master' into vapopov/heartbeat-local-time-snap

vapopov committed:
Fix test

vapopov committed:
Drop timeReconciliation in order to use direct handler functions

vapopov committed:
Merge remote-tracking branch 'origin/master' into vapopov/heartbeat-local-time-snap

fspmarshall reviewed (approved):
LGTM once ping interval has tweaked to be a bit less onerous at scale.

vapopov committed:
Make time reconciliation with variable duration

espadolini commented on lib/auth/server_monitor.go:
```diff
@@ -0,0 +1,161 @@
+		handle.VisitInstanceState(func(ref inventory.InstanceStateRef) (update inventory.InstanceStateUpdate) {
+			if ref.LastHeartbeat != nil && ref.LastHeartbeat.GetLastMeasurement() != nil {
+				m := ref.LastHeartbeat.GetLastMeasurement()
+				diff := m.ControllerSystemClock.Sub(m.SystemClock) - m.RequestDuration/2
```
The `SystemClock` is not going to be available until agents are updated, won't this report a difference for every agent that's not responding with the local time to the ping?

vapopov replied:
For nodes without `InstanceSpecV1.LastMeasurement` we should just ignore check, so we skip them for report

espadolini replied:
Where's the logic that omits LastMeasurement if the ping response doesn't include the instance system clock?

vapopov replied:
got you, I've added here https://github.com/gravitational/teleport/pull/44489/commits/ebbbde65c81cb7e3de69ac19da7d57aa9cdfbf2d

espadolini commented on lib/auth/server_monitor.go:
```diff
@@ -0,0 +1,161 @@
+		handle.VisitInstanceState(func(ref inventory.InstanceStateRef) (update inventory.InstanceStateUpdate) {
+			if ref.LastHeartbeat != nil && ref.LastHeartbeat.GetLastMeasurement() != nil {
+				m := ref.LastHeartbeat.GetLastMeasurement()
+				diff := m.ControllerSystemClock.Sub(m.SystemClock) - m.RequestDuration/2
```
This might deserve a little comment explaining why we're subtracting half of the RTT.

espadolini commented on lib/inventory/controller.go:
```diff
@@ -288,11 +320,15 @@ func (c *Controller) RegisterControlStream(stream client.UpstreamInventoryContro
+	}, interval.SubInterval[intervalKey]{
+		Key:      instanceTimeReconciliation,
+		Duration: c.timeReconciliationInterval,
+		Jitter:   seventhJitter,
```
That's not really going to guarantee that the first ping is done by the time we do the first instance heartbeat keepalive - IMO waiting 3 more minutes is not really a problem, since we have to handle outdated agents not sending us the local time anyway.

espadolini replied:
Well, it will initiate the ping but we have no guarantee that we'll actually receive the response before the heartbeat hits, especially if we rolled very low on `firstDuration`. This is why we should make sure that we won't report anything about a server if we haven't received a timestamp yet.

vapopov replied:
ah, you mean that response of the ping must be after heart, we should set `LastMeasurement = nil` in such case https://github.com/gravitational/teleport/pull/44489/commits/ebbbde65c81cb7e3de69ac19da7d57aa9cdfbf2d
so it must be ignored in general check

espadolini reviewed (commented):
Heads up, now that #49433 landed you're gonna have to change the subinterval into a `delay.Delay` and move the sending logic one level higher. Is time drift monitoring planned to be backported to v16, or only v17?

vapopov committed:
Replace with duration type in proto
Add test cleanup for possible goroutine leak
Add comment about half request duration

vapopov committed:
Merge remote-tracking branch 'origin/master' into vapopov/heartbeat-local-time-snap

vapopov commented on lib/auth/server_monitor.go:
```diff
@@ -0,0 +1,161 @@
+	message := "Incorrect system clock detected in the cluster, which may lead to certificate validation issues.\n" +
+		"Ensure that the clock is accurate on all nodes to avoid potential access problems.\n" +
+		"All comparisons are made with the Auth service system clock." +
+		"List of servers with a time drift: \n" + strings.Join(messages, "\n")
```
```bash
root@3e255c82fc04:/go/teleport# ./build/tctl notifications ls --all
ID                                   Created             Expires             Title                                                                                                                                                                                                                                                                                                                 Labels
------------------------------------ ------------------- ------------------- --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ------
cluster-monitor-system-clock-warning 27 Nov 24 23:36 UTC 10 Dec 24 11:36 UTC Incorrect system clock detected in the cluster, which may lead to certificate validation issues.
Ensure that the clock is accurate on all nodes to avoid potential access problems.
All comparisons are made with the Auth service system clock.List of servers with a time drift:
 - test-id[Node] is 10h0m0s ahead
```

<img width="487" alt="image" src="https://github.com/user-attachments/assets/710808f8-8b5f-4657-860e-c516e525012e">

<img width="1512" alt="image" src="https://github.com/user-attachments/assets/0da5b7cd-5820-4d8d-ae56-cb06139bcb4f">

looks like for the web, for spacing we should use something different

vapopov replied:
I made some adjustments, moved message to `ContentLabel` and changed format to be readable inline

```bash
root@3e255c82fc04:/go/teleport# ./build/tctl notifications ls --all
ID                                   Created             Expires             Title                                          Labels
------------------------------------ ------------------- ------------------- ---------------------------------------------- ------
cluster-monitor-system-clock-warning 28 Nov 24 01:28 UTC 28 Dec 24 01:28 UTC Incorrect system clock detected in the cluster
```

<img width="488" alt="image" src="https://github.com/user-attachments/assets/2d67d08a-caf4-4bc7-96ac-0e48fb0f6968">
<img width="1512" alt="image" src="https://github.com/user-attachments/assets/4628d485-e4ea-49bb-a56a-b94969ee6975">

vapopov replied:
notifications as far as I understand aren't shown with `tsh status` and `tsh login` only cluster alerts, so you need to request them manually `tctl notifications ls --all`

vapopov committed:
Format notification message

vapopov committed:
Add UpstreamInventoryPong.SystemClock check
